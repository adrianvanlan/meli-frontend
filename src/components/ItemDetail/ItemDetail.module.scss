@import '/src/styles/base';

.ItemDetail {
  text-align: left;
  margin-bottom: 100px;

  &__Wrapper {
    margin: 0 auto;
    background-color: #fff;
    padding: 32px;
  }

  &__Content {
    display: flex;
    justify-content: space-between;
    flex-direction: column;

    &__Left {
      width: 100%;
      display: flex;
      flex-direction: column;
      flex-wrap: wrap;
    }

    &__Image {
      width: 100%;
      overflow: hidden;
      min-height: 200px;

      img {
        max-width: 680px;
        width: 100%;
        height: auto;
        max-height: 700px;
        object-fit: contain;
      }
    }
    &__Description {
      padding: 96px 0 0;
      width: 100%;

      &__Title {
        font-size: 28px;
      }

      &__Text {
        padding: 32px 0 0;
        font-size: 16px;
        font-weight: 400;
        color: $secondary-text;
        line-height: 27px;
        word-break: break-word;
      }
    }

    &__Right {
      width: 100%;
      display: flex;
      flex-direction: column;
      flex-wrap: wrap;

      &__Info {
        color: $primary-text;
        background-color: $white;
        font-size: 14px;

        &__Extra {
          color: $secondary-text;
          margin-top: 10px;
        }

        &__Title {
          font-size: 24px;
          font-weight: 500;
          margin-top: 16px;
          line-height: 1.2;
        }

        &__Price {
          font-size: 46px;
          font-weight: 400;
          margin-top: 32px;
          display: flex;

          span {
            line-height: 0.9;
          }

          &__Decimals {
            font-size: 20px;
            font-weight: 400;
            align-self: flex-start;
            margin-left: 5px;
          }
        }

        &__Action {
          margin-top: 60px;

          button {
            background-color: $primary-link;
            border: none;
            border-radius: $border-radius;
            color: #fff;
            cursor: pointer;
            font-size: 18px;
            font-weight: 400;
            padding: 0px 24px;
            width: 100%;
            transition: all 0.2s ease-in;
            height: 48px;

            &:focus {
              outline: none;
            }

            &:hover {
              background-color: $primary-link-hover;
            }
          }
        }
      }
    }
  }
}

@media (min-width: 1024px) {
  .ItemDetail {
    &__Content {
      flex-direction: row;

      &__Left {
        width: 60%;
      }

      &__Right {
        width: 325px;

        &__Info {
          position: sticky;
          top: 30px;
        }
      }
    }
  }
}
