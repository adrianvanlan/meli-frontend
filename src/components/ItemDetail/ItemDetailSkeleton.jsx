import Skeleton from 'react-loading-skeleton';
import styles from './ItemDetail.module.scss';

function ItemDetailSkeleton() {
  return (
    <div className={styles.ItemDetail}>
      <div className={styles.ItemDetail__Breadcrumb}>
        <Skeleton width={280} height={20} style={{ margin: '18px 0' }} />
      </div>
      <div className={styles.ItemDetail__Wrapper}>
        <div className={styles.ItemDetail__Content}>
          <div className={styles.ItemDetail__Content__Left}>
            <div className={styles.ItemDetail__Content__Image}>
              <Skeleton height={700} width="100%" />
            </div>
          </div>
          <div className={styles.ItemDetail__Content__Right}>
            <div className={styles.ItemDetail__Content__Right__Info}>
              <div className={styles.ItemDetail__Content__Right__Info__Extra}>
                <Skeleton height={10} width={150} />
              </div>
              <h1 className={styles.ItemDetail__Content__Right__Info__Title}>
                <Skeleton count={2} height={20} width={'80%'} />
              </h1>
              <div className={styles.ItemDetail__Content__Right__Info__Price}>
                <Skeleton height={40} width={175} />
              </div>
              <div className={styles.ItemDetail__Content__Right__Info__Action}>
                <Skeleton height={50} width={'100%'} />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

export default ItemDetailSkeleton;
