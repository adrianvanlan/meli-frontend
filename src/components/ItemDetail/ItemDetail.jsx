import { useEffect, useState } from 'react';

import Breadcrumb from '../Breadcrumb/Breadcrumb';
import { Helmet } from 'react-helmet-async';
import ItemDetailSkeleton from './ItemDetailSkeleton';
import Skeleton from 'react-loading-skeleton';
import { getCurrencyPrice } from '../../helpers';
import styles from './ItemDetail.module.scss';
import { useItem } from '../../hooks/useItems';
import { useParams } from 'react-router-dom';

function ItemDetail() {
  const params = useParams({});
  const [itemId, setItemId] = useState(null);
  const { item: itemData, isLoading } = useItem(itemId);

  const renderDescription = (data) => {
    if (data && data.description !== '') {
      return data.description;
    } else {
      return 'Anuncio sin descripciÃ³n';
    }
  };

  useEffect(() => {
    const { id } = params;
    if (id) {
      setItemId(id);
    }
  }, [params]);

  useEffect(() => {
    window.scroll(0, 0);
  }, []);

  if (isLoading) {
    return <Skeleton wrapper={ItemDetailSkeleton} />;
  }

  return (
    <div className={styles.ItemDetail}>
      {itemData && itemData.title && (
        <Helmet>
          <title>{`${itemData.title} | Mercado Libre ðŸ“¦`}</title>
          <meta
            name="description"
            content={`EnvÃ­os gratis en el dÃ­a âœ“ ComprÃ¡ online de manera segura con Compra Protegida ${itemData.title}`}
          />
        </Helmet>
      )}

      {itemData && itemData.categories && (
        <div className={styles.ItemDetail__Breadcrumb}>
          <Breadcrumb categories={itemData.categories} />
        </div>
      )}
      <div className={styles.ItemDetail__Wrapper}>
        <div className={styles.ItemDetail__Content}>
          <div className={styles.ItemDetail__Content__Left}>
            <div className={styles.ItemDetail__Content__Image}>
              {itemData && itemData.picture && (
                <img src={itemData.picture} alt={`Imagen ${itemData.title}`} />
              )}
            </div>
          </div>
          <div className={styles.ItemDetail__Content__Right}>
            <div className={styles.ItemDetail__Content__Right__Info}>
              <div className={styles.ItemDetail__Content__Right__Info__Extra}>
                {itemData && (itemData.sold_quantity || itemData.condition) && (
                  <span>
                    {itemData.condition === 'new' ? 'Nuevo' : 'Usado'}
                    {itemData && itemData.sold_quantity > 0
                      ? ` - ${itemData.sold_quantity} vendido${itemData.sold_quantity === 1 ? '' : 's'}`
                      : null}
                  </span>
                )}
              </div>
              <h1 className={styles.ItemDetail__Content__Right__Info__Title}>
                {itemData && itemData.title && itemData.title}
              </h1>
              <div className={styles.ItemDetail__Content__Right__Info__Price}>
                {itemData && itemData.price && (
                  <span>
                    {getCurrencyPrice(
                      itemData.price.amount,
                      itemData.price.currency,
                    )}
                  </span>
                )}
                {itemData.price && itemData.price.decimals > 0 && (
                  <span
                    className={
                      styles.ItemDetail__Content__Right__Info__Price__Decimals
                    }
                  >
                    {itemData.price.decimals}
                  </span>
                )}
              </div>

              <div className={styles.ItemDetail__Content__Right__Info__Action}>
                <button type="button">Comprar</button>
              </div>
            </div>
          </div>
        </div>
        <div className={styles.ItemDetail__Content__Description}>
          <div className={styles.ItemDetail__Content__Description__Title}>
            DescripciÃ³n del Producto
          </div>
          <div className={styles.ItemDetail__Content__Description__Text}>
            {renderDescription(itemData)}
          </div>
        </div>
      </div>
    </div>
  );
}

export default ItemDetail;
