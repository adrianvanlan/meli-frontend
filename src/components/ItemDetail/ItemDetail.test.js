import ItemDetail from './ItemDetail';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';

jest.mock('react-helmet-async', () => ({
  Helmet: () => jest.fn(),
  HelmetProvider: () => jest.fn(),
}));

jest.mock('../../hooks/useItems', () => ({
  useItem: () => ({
    item: {
      id: 'MLA123456',
      title: 'Item Title',
      price: {
        currency: 'ARS',
        amount: 123,
        decimals: 0,
      },
      picture: '',
      condition: 'new',
      free_shipping: true,
      sold_quantity: 1,
      description: 'Item Description',
      categories: ['Category 1', 'Category 2'],
    },
  }),
}));

it('expect to render ItemDetail component', () => {
  const wrapper = render(
    <MemoryRouter>
      <ItemDetail />
    </MemoryRouter>,
  );

  expect(wrapper).toMatchSnapshot();
});
