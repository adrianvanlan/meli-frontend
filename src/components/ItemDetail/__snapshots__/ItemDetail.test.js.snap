// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render ItemDetail component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ItemDetail"
      >
        <div
          class="ItemDetail__Breadcrumb"
        >
          <ul
            class="Breadcrumb"
          >
            <li
              class="Breadcrumb__Item"
            >
              <span>
                Category 1
              </span>
            </li>
            <li
              class="Breadcrumb__Item"
            >
              <svg>
                arrowRight.svg
              </svg>
              <span>
                Category 2
              </span>
            </li>
          </ul>
        </div>
        <div
          class="ItemDetail__Wrapper"
        >
          <div
            class="ItemDetail__Content"
          >
            <div
              class="ItemDetail__Content__Left"
            >
              <div
                class="ItemDetail__Content__Image"
              />
            </div>
            <div
              class="ItemDetail__Content__Right"
            >
              <div
                class="ItemDetail__Content__Right__Info"
              >
                <div
                  class="ItemDetail__Content__Right__Info__Extra"
                >
                  <span>
                    Nuevo
                     - 1 vendido
                  </span>
                </div>
                <h1
                  class="ItemDetail__Content__Right__Info__Title"
                >
                  Item Title
                </h1>
                <div
                  class="ItemDetail__Content__Right__Info__Price"
                >
                  <span>
                    $ 123
                  </span>
                </div>
                <div
                  class="ItemDetail__Content__Right__Info__Action"
                >
                  <button
                    type="button"
                  >
                    Comprar
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ItemDetail__Content__Description"
          >
            <div
              class="ItemDetail__Content__Description__Title"
            >
              Descripción del Producto
            </div>
            <div
              class="ItemDetail__Content__Description__Text"
            >
              Item Description
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ItemDetail"
    >
      <div
        class="ItemDetail__Breadcrumb"
      >
        <ul
          class="Breadcrumb"
        >
          <li
            class="Breadcrumb__Item"
          >
            <span>
              Category 1
            </span>
          </li>
          <li
            class="Breadcrumb__Item"
          >
            <svg>
              arrowRight.svg
            </svg>
            <span>
              Category 2
            </span>
          </li>
        </ul>
      </div>
      <div
        class="ItemDetail__Wrapper"
      >
        <div
          class="ItemDetail__Content"
        >
          <div
            class="ItemDetail__Content__Left"
          >
            <div
              class="ItemDetail__Content__Image"
            />
          </div>
          <div
            class="ItemDetail__Content__Right"
          >
            <div
              class="ItemDetail__Content__Right__Info"
            >
              <div
                class="ItemDetail__Content__Right__Info__Extra"
              >
                <span>
                  Nuevo
                   - 1 vendido
                </span>
              </div>
              <h1
                class="ItemDetail__Content__Right__Info__Title"
              >
                Item Title
              </h1>
              <div
                class="ItemDetail__Content__Right__Info__Price"
              >
                <span>
                  $ 123
                </span>
              </div>
              <div
                class="ItemDetail__Content__Right__Info__Action"
              >
                <button
                  type="button"
                >
                  Comprar
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ItemDetail__Content__Description"
        >
          <div
            class="ItemDetail__Content__Description__Title"
          >
            Descripción del Producto
          </div>
          <div
            class="ItemDetail__Content__Description__Text"
          >
            Item Description
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
